### DISCUSSION TITLE
CustomNormalizer implementation concern

### DISCUSSION THREAD

**Discussion by rmikalkenas** [symfony/symfony#43768] [Q&A] [answered]
Symfony serializer has a [CustomNormalizer](https://github.com/symfony/serializer/blob/5.3/Normalizer/CustomNormalizer.php) implementation, which gives the possibility to define custom normalization and denormalization flow of object by implementing NormalizableInterface and DenormalizableInterface respectively. 

My main concern is with `denormalize` method implementation and DenormalizableInterface contract. Based on it there is no way to implement this interface if object has a constructor, because of [current logic](https://github.com/symfony/serializer/blob/5d7f068253ac3e7c62964ebdda491b06d401059a/Normalizer/CustomNormalizer.php#L38) 

From my gut feeling DenormalizableInterface should have a static denormalize method with a return value. In that way the object who is implementing this interface would be responsible for creating itself (and not CustomNormalizer) and denormalizing its state.

Would be interesting to hear others opinion on this case, maybe I'm missing something and such implementation was done on purpose

**Comment by lyrixx** [comment, marked as answer]
You can pass the object to the `denormalize()` method : 

```php
$this->serializer->denormalize($rawData, YouObject::class, context: [
  AbstractNormalizer::OBJECT_TO_POPULATE => new YouObject($arg1, $args2, $arg3),
]);
```

**Reply by rmikalkenas** [reply]
@lyrixx I know that, but then whats the point of that DenormalizableInterface if you have to denormalize an object before passing to serializer..? 
I see a couple of drawbacks with this approach:
1. There is not enough to implement DenormalizableInterface, in addition you have to add additional logic for passing object to populate attribute. Which means that object deconstruction process is scattered in multiple places (no single point of whole logic)
2. Secondly if I would like to implement such interface for object which Im using for async envelopes (symfony messenger) that would mean I have to rewrite messenger's serializer to adapt logic based in object to populate attribute which from my point of view is a huge overhead.

What do you think about it? 

Also would be interesting to hear @Seldaek opinion as he is the owner of this denormalizer (at least annotation is saying that)

**Reply by Seldaek** [reply]
Sorry but I haven't used this code in far too long to have any kind of valid opinion on this right now.

**Reply by lyrixx** [reply]
> I know that, but then whats the point of that DenormalizableInterface if you have to denormalize an object before passing to serializer..?

not denormalize it, only intanciate it..

But I agree with you, a new interface can be added, or updating the current one (but the BC must be preserved)

**Reply by rmikalkenas** [reply]
Ok, will think of possible new interface as making a BC for the existing one might be impossible 😅
